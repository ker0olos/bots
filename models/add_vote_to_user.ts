import {
  BooleanExpr,
  Client,
  fql,
  InventoryExpr,
  NumberExpr,
  ResponseExpr,
  UserExpr,
} from './fql.ts';

import {
  getGuild,
  getInstance,
  getInventory,
  getUser,
  Inventory,
  rechargePulls,
  User,
} from './get_user_inventory.ts';

export function addGuarantee(
  {
    user,
    guarantee,
  }: {
    user: UserExpr;
    guarantee: NumberExpr;
  },
): ResponseExpr {
  return fql.Let(
    {
      // increased by factor of 3
      cost: fql.If(
        fql.Equals(guarantee, 3),
        4,
        fql.If(
          fql.Equals(guarantee, 4),
          12,
          fql.If(
            fql.Equals(guarantee, 5),
            36,
            fql.Null(),
          ),
        ),
      ),
    },
    ({ cost }) =>
      fql.If(
        fql.IsNull(cost),
        { ok: false, error: 'NOT_SUPPORTED' },
        fql.If(
          fql.GTE(fql.Select(['data', 'availableVotes'], user, 0), cost),
          fql.Let(
            {
              updatedUser: fql.Update<User>(fql.Ref(user), {
                availableVotes: fql.Subtract(
                  fql.Select(['data', 'availableVotes'], user),
                  cost,
                ),
                guarantees: fql.Append(
                  guarantee,
                  fql.Select(['data', 'guarantees'], user, []),
                ),
              }),
            },
            ({ updatedUser }) =>
              ({
                ok: true,
                user: fql.Ref(updatedUser),
              }) as unknown as ResponseExpr,
          ),
          { ok: false, error: 'INSUFFICIENT_VOTES' },
        ),
      ),
  );
}

export function addPulls(
  { user, inventory, votes }: {
    user: UserExpr;
    inventory: InventoryExpr;
    votes: NumberExpr;
  },
): ResponseExpr {
  return fql.If(
    fql.GTE(fql.Select(['data', 'availableVotes'], user, 0), votes),
    fql.Let(
      {
        updatedUser: fql.Update<User>(fql.Ref(user), {
          availableVotes: fql.Subtract(
            fql.Select(['data', 'availableVotes'], user),
            votes,
          ),
        }),
        updatedInventory: fql.Update<Inventory>(fql.Ref(inventory), {
          availablePulls: fql.Add(
            fql.Select(['data', 'availablePulls'], inventory),
            votes,
          ),
        }),
      },
      ({ updatedInventory }) =>
        ({
          ok: true,
          inventory: fql.Ref(updatedInventory),
        }) as unknown as ResponseExpr,
    ),
    { ok: false, error: 'INSUFFICIENT_VOTES' },
  );
}

export function addVote(
  { user, weekend }: { user: UserExpr; weekend: BooleanExpr },
): ResponseExpr {
  return fql.Let({
    user: fql.Update<User>(fql.Ref(user), {
      lastVote: fql.Now(),
      totalVotes: fql.Add(
        fql.Select(['data', 'totalVotes'], user, 0),
        1,
      ),
      availableVotes: fql.Add(
        fql.Select(['data', 'availableVotes'], user, 0),
        // double votes on weekends
        fql.If(fql.Equals(weekend, true), 2, 1),
      ),
    }),
  }, () => ({ ok: true }) as unknown as ResponseExpr);
}

export default function (client: Client): {
  indexers?: (() => Promise<void>)[];
  resolvers?: (() => Promise<void>)[];
} {
  return {
    resolvers: [
      fql.Resolver({
        client,
        name: 'exchange_votes_for_guarantees',
        lambda: (userId: string, guarantee: number) => {
          return fql.Let(
            { user: getUser(userId) },
            ({ user }) => addGuarantee({ user, guarantee }),
          );
        },
      }),
      fql.Resolver({
        client,
        name: 'exchange_votes_for_pulls',
        lambda: (userId: string, guildId: string, votes: number) => {
          return fql.Let(
            {
              user: getUser(userId),
              guild: getGuild(guildId),
              instance: getInstance(fql.Var('guild')),
              _inventory: getInventory({
                user: fql.Var('user'),
                instance: fql.Var('instance'),
              }),
              inventory: rechargePulls({
                inventory: fql.Var('_inventory'),
              }),
            },
            ({ user, inventory }) => addPulls({ user, inventory, votes }),
          );
        },
      }),
      fql.Resolver({
        client,
        name: 'add_vote_to_user',
        lambda: (userId: string, weekend: boolean) => {
          return fql.Let({ user: getUser(userId), weekend }, addVote);
        },
      }),
    ],
  };
}
