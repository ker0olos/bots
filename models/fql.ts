// deno-lint-ignore-file no-explicit-any

import {
  Client,
  query,
} from 'https://deno.land/x/fauna@5.0.0-deno-alpha9/mod.js';

import {
  Expr,
  ExprArg,
  type query as _query,
} from 'https://deno.land/x/fauna@5.0.0-deno-alpha9/mod.d.ts';

const _fql = query as typeof _query;

type TypedExpr<T> = Expr & { type?: T };

export type StringExpr = string | TypedExpr<'string'>;
export type NumberExpr = number | TypedExpr<'number'>;
export type BooleanExpr = boolean | TypedExpr<'boolean'>;

export type NullExpr = TypedExpr<'null'>;
export type TimeExpr = TypedExpr<'time'>;
export type IndexExpr = TypedExpr<'index'>;
export type MatchExpr = TypedExpr<'match'>;
export type RefExpr = TypedExpr<'ref'>;

export type ResponseExpr = TypedExpr<'response'>;

export type UserExpr = TypedExpr<'user'>;
export type GuildExpr = TypedExpr<'guild'>;
export type InstanceExpr = TypedExpr<'instance'>;
export type InventoryExpr = TypedExpr<'inventory'>;
export type CharacterExpr = TypedExpr<'character'>;
export type PackExpr = TypedExpr<'pack'>;
export type ManifestExpr = TypedExpr<'manifest'>;

function Ref(document: Expr): RefExpr {
  return _fql.Select('ref', document);
}

function Id(name: StringExpr, id: StringExpr): RefExpr {
  return _fql.Ref(_fql.Collection(name), id);
}

function Create<T = Expr>(collection: string, data: T): Expr {
  return _fql.Create(_fql.Collection(collection), {
    data,
  });
}

function Update<T = Expr>(ref: RefExpr, data: Partial<T>): Expr {
  return _fql.Update(ref, {
    data,
  });
}

function Index(name: string): IndexExpr {
  return _fql.FaunaIndex(name);
}

function Match(index: IndexExpr, ...terms: ExprArg[]): MatchExpr {
  return _fql.Match(index, ...terms);
}

function Get(refOrMatch: RefExpr | MatchExpr): Expr {
  return _fql.Get(refOrMatch);
}

function Append(ref: RefExpr, items: Expr): RefExpr[] {
  return _fql.Append(ref, items) as unknown as RefExpr[];
}

function Remove(ref: RefExpr, items: Expr): RefExpr[] {
  return _fql.Difference(items, [ref]) as unknown as RefExpr[];
}

function Includes(value: ExprArg, documentOrArray: Expr): BooleanExpr {
  return _fql.ContainsValue(value, documentOrArray) as unknown as BooleanExpr;
}
function Paginate(
  expr: Expr,
  { size, before, after }: { size?: number; before?: any; after?: any },
): Expr {
  return _fql.Paginate(expr, {
    size,
    before,
    after,
  });
}

// function Map<T = Expr>(
//   refOrMatch: RefExpr | MatchExpr,
//   map: (...args: ExprArg[]) => Expr,
// ): T[] {
//   return _fql.Map(refOrMatch, map) as unknown as T[];
// }

function IsNonEmpty(expr: Expr): BooleanExpr {
  return _fql.IsNonEmpty(expr);
}

function Min(a: NumberExpr, b: NumberExpr): NumberExpr {
  return _fql.Min(a, b);
}

function GTE(a: NumberExpr, b: NumberExpr): BooleanExpr {
  return _fql.GTE(a, b);
}

function LTE(a: NumberExpr, b: NumberExpr): BooleanExpr {
  return _fql.LTE(a, b);
}

function Multiply(a: NumberExpr, b: NumberExpr): NumberExpr {
  return _fql.Multiply(a, b);
}

function Divide(a: NumberExpr, b: NumberExpr): NumberExpr {
  return _fql.Divide(a, b);
}

function Subtract(a: NumberExpr, b: NumberExpr): NumberExpr {
  return _fql.Subtract(a, b);
}

function Add(a: NumberExpr, b: NumberExpr): NumberExpr {
  return _fql.Add(a, b);
}

function If<A = Expr, B = Expr>(
  cond: BooleanExpr,
  _then: Expr | ExprArg,
  _else: Expr | ExprArg,
): A | B {
  return _fql.If(cond, _then, _else) as A | B;
}

function Equals<A = Expr, B = Expr>(a?: A, b?: B): BooleanExpr {
  // deno-lint-ignore no-non-null-assertion
  return _fql.Equals(a!, b!);
}

function Concat(s: StringExpr[], sep?: StringExpr): StringExpr {
  return _fql.Concat(s, sep ?? '');
}

function ToString(n: NumberExpr): StringExpr {
  return _fql.ToString(n);
}

function IsNull(expr?: ExprArg): BooleanExpr {
  // deno-lint-ignore no-non-null-assertion
  return _fql.Equals(expr!, fql.Null());
}

function Let<T, U>(
  params: { [K in keyof T]: T[K] },
  cb: (shadows: typeof params) => U,
): U {
  const shadows: any = {};

  for (const obj of Object.keys(params)) {
    shadows[obj] = _fql.Var(obj);
  }

  return _fql.Let(params, cb(shadows) as any) as U;
}

function Var(name: StringExpr): Expr {
  return _fql.Var(name);
}

function Select(
  path: (StringExpr | NumberExpr)[],
  from: Expr,
  _default?: any,
): Expr {
  return _fql.Select(path, from, _default);
}

function Merge(
  obj1: Record<string, any>,
  obj2: Record<string, any>,
): Expr {
  return _fql.Merge(obj1, obj2);
}

function Reverse(expr: Expr): Expr {
  return _fql.Reverse(expr);
}

function TimeDiffInMinutes(a: TimeExpr, b: TimeExpr): NumberExpr {
  return _fql.TimeDiff(a, b, 'minutes');
}

// function And(a: BooleanExpr, b: BooleanExpr): BooleanExpr {
//   return _fql.And(a, b);
// }

// function Or(a: BooleanExpr, b: BooleanExpr): BooleanExpr {
//   return _fql.Or(a, b);
// }

function TimeAddInMinutes(t: TimeExpr, offset: NumberExpr): NumberExpr {
  return _fql.TimeAdd(t, offset, 'minutes');
}

function Now(): TimeExpr {
  return _fql.Now();
}

function Null(): NullExpr {
  return null as unknown as NullExpr;
}

function Indexer(
  { client, name, unique, collection, values, terms }: {
    client: Client;
    unique: boolean;
    collection: string;
    name: string;
    terms?: {
      field: string[];
    }[];
    values?: {
      field: string[];
      reverse?: boolean;
    }[];
  },
): () => Promise<void> {
  const params = {
    name,
    unique,
    source: _fql.Collection(collection),
    terms,
    values,
  };

  return () =>
    client.query(
      _fql.If(
        _fql.Exists(_fql.FaunaIndex(name)),
        _fql.Update(_fql.FaunaIndex(name), params),
        _fql.CreateIndex(params),
      ),
    );
}

function Resolver(
  { client, name, lambda }: {
    name: string;
    client: Client;
    lambda: (...vars: any[]) => ExprArg;
  },
): () => Promise<void> {
  const params = {
    name,
    body: _fql.Query(lambda),
  };

  return () =>
    client.query(
      _fql.If(
        _fql.Exists(_fql.FaunaFunction(name)),
        _fql.Update(_fql.FaunaFunction(name), params),
        _fql.CreateFunction(params),
      ),
    );
}

export const fql = {
  // And,
  // Map,
  // Or,
  Add,
  Append,
  Concat,
  Create,
  Divide,
  Equals,
  Get,
  GTE,
  Id,
  If,
  Includes,
  Index,
  Indexer,
  IsNonEmpty,
  IsNull,
  Let,
  LTE,
  Match,
  Merge,
  Min,
  Multiply,
  Now,
  Null,
  Paginate,
  Ref,
  Remove,
  Resolver,
  Reverse,
  Select,
  Subtract,
  TimeAddInMinutes,
  TimeDiffInMinutes,
  ToString,
  Update,
  Var,
};

export type { Client };
