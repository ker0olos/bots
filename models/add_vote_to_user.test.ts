// deno-lint-ignore-file no-explicit-any

import { assertEquals } from 'https://deno.land/std@0.179.0/testing/asserts.ts';

import {
  assertSpyCall,
  assertSpyCalls,
  returnsNext,
  stub,
} from 'https://deno.land/std@0.179.0/testing/mock.ts';

import { assertSnapshot } from 'https://deno.land/std@0.179.0/testing/snapshot.ts';

import {
  FakeAdd,
  FakeClient,
  FakeEquals,
  FakeIf,
  FakeLet,
  FakeNow,
  FakeRef,
  FakeUpdate,
} from './fql.mock.ts';

import { fql } from './fql.ts';

import { addVote, default as Model } from './add_vote_to_user.ts';

Deno.test('voting', async (test) => {
  await test.step('normal', () => {
    const date = new Date();

    const letStub = FakeLet();
    const refStub = FakeRef();
    const ifStub = FakeIf();
    const equalsStub = FakeEquals();
    const addStub = FakeAdd();

    const nowStub = FakeNow(date);

    const selectStub = stub(
      fql,
      'Select',
      returnsNext([
        5,
        1,
      ]) as any,
    );

    const updateStub = FakeUpdate();

    try {
      const result = addVote({ user: 'user' as any, weekend: false }) as any;

      assertSpyCall(ifStub, 0, {
        args: [false, 2, 1],
        returned: 1,
      });

      assertSpyCall(selectStub, 0, {
        args: [
          [
            'data',
            'totalVotes',
          ],
          'user' as any,
          0,
        ],
      });

      assertSpyCall(selectStub, 1, {
        args: [
          [
            'data',
            'availableVotes',
          ],
          'user' as any,
          0,
        ],
      });

      assertSpyCall(addStub, 0, {
        args: [5, 1],
      });

      assertSpyCall(addStub, 1, {
        args: [1, 1],
      });

      assertSpyCall(updateStub, 0, {
        args: [{ ref: 'user' } as any, {
          lastVote: date,
          totalVotes: 6,
          availableVotes: 2,
        }],
      });

      assertEquals(result, {
        ok: true,
      });
    } finally {
      letStub.restore();
      refStub.restore();
      ifStub.restore();
      equalsStub.restore();
      addStub.restore();
      nowStub.restore();
      selectStub.restore();
      updateStub.restore();
    }
  });

  await test.step('weekend', () => {
    const date = new Date();

    const letStub = FakeLet();
    const refStub = FakeRef();
    const ifStub = FakeIf();
    const equalsStub = FakeEquals();
    const addStub = FakeAdd();

    const nowStub = FakeNow(date);

    const selectStub = stub(
      fql,
      'Select',
      returnsNext([
        5,
        1,
      ]) as any,
    );

    const updateStub = FakeUpdate();

    try {
      const result = addVote({ user: 'user' as any, weekend: true }) as any;

      assertSpyCall(ifStub, 0, {
        args: [true, 2, 1],
        returned: 2,
      });

      assertSpyCall(selectStub, 0, {
        args: [
          [
            'data',
            'totalVotes',
          ],
          'user' as any,
          0,
        ],
      });

      assertSpyCall(selectStub, 1, {
        args: [
          [
            'data',
            'availableVotes',
          ],
          'user' as any,
          0,
        ],
      });

      assertSpyCall(addStub, 0, {
        args: [5, 1],
      });

      assertSpyCall(addStub, 1, {
        args: [1, 2],
      });

      assertSpyCall(updateStub, 0, {
        args: [{ ref: 'user' } as any, {
          lastVote: date,
          totalVotes: 6,
          availableVotes: 3,
        }],
      });

      assertEquals(result, {
        ok: true,
      });
    } finally {
      letStub.restore();
      refStub.restore();
      ifStub.restore();
      equalsStub.restore();
      addStub.restore();
      nowStub.restore();
      selectStub.restore();
      updateStub.restore();
    }
  });
});

Deno.test('model', async (test) => {
  const client = FakeClient();

  Model(client as any).indexers?.forEach((q) => q());
  Model(client as any).resolvers?.forEach((q) => q());

  assertSpyCalls(client.query, 1);

  await assertSnapshot(test, client.query.calls[0].args);
});
