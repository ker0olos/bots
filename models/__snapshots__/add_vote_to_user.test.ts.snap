export const snapshot = {};

snapshot[`model 1`] = `
[
  Expr {
    raw: {
      else: Expr {
        raw: {
          create_function: Expr {
            raw: {
              object: {
                body: Expr {
                  raw: {
                    query: Expr {
                      raw: {
                        expr: Expr {
                          raw: {
                            in: Expr {
                              raw: {
                                in: Expr {
                                  raw: {
                                    object: {
                                      ok: true,
                                    },
                                  },
                                },
                                let: [
                                  {
                                    user: Expr {
                                      raw: {
                                        params: Expr {
                                          raw: {
                                            object: {
                                              data: Expr {
                                                raw: {
                                                  object: {
                                                    availableVotes: Expr {
                                                      raw: {
                                                        add: Expr {
                                                          raw: [
                                                            Expr {
                                                              raw: {
                                                                default: 0,
                                                                from: Expr {
                                                                  raw: {
                                                                    var: "user",
                                                                  },
                                                                },
                                                                select: Expr {
                                                                  raw: [
                                                                    "data",
                                                                    "availableVotes",
                                                                  ],
                                                                },
                                                              },
                                                            },
                                                            Expr {
                                                              raw: {
                                                                else: 1,
                                                                if: Expr {
                                                                  raw: {
                                                                    equals: Expr {
                                                                      raw: [
                                                                        Expr {
                                                                          raw: {
                                                                            var: "weekend",
                                                                          },
                                                                        },
                                                                        true,
                                                                      ],
                                                                    },
                                                                  },
                                                                },
                                                                then: 2,
                                                              },
                                                            },
                                                          ],
                                                        },
                                                      },
                                                    },
                                                    lastVote: Expr {
                                                      raw: {
                                                        now: null,
                                                      },
                                                    },
                                                    totalVotes: Expr {
                                                      raw: {
                                                        add: Expr {
                                                          raw: [
                                                            Expr {
                                                              raw: {
                                                                default: 0,
                                                                from: Expr {
                                                                  raw: {
                                                                    var: "user",
                                                                  },
                                                                },
                                                                select: Expr {
                                                                  raw: [
                                                                    "data",
                                                                    "totalVotes",
                                                                  ],
                                                                },
                                                              },
                                                            },
                                                            1,
                                                          ],
                                                        },
                                                      },
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                        update: Expr {
                                          raw: {
                                            from: Expr {
                                              raw: {
                                                var: "user",
                                              },
                                            },
                                            select: "ref",
                                          },
                                        },
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                            let: [
                              {
                                user: Expr {
                                  raw: {
                                    in: Expr {
                                      raw: {
                                        else: Expr {
                                          raw: {
                                            create: Expr {
                                              raw: {
                                                collection: "user",
                                              },
                                            },
                                            params: Expr {
                                              raw: {
                                                object: {
                                                  data: Expr {
                                                    raw: {
                                                      object: {
                                                        badges: Expr {
                                                          raw: [
                                                            Expr {
                                                              raw: {
                                                                id: "357600097731608662",
                                                                ref: Expr {
                                                                  raw: {
                                                                    collection: "badge",
                                                                  },
                                                                },
                                                              },
                                                            },
                                                          ],
                                                        },
                                                        id: Expr {
                                                          raw: {
                                                            var: "userId",
                                                          },
                                                        },
                                                        inventories: Expr {
                                                          raw: [
                                                          ],
                                                        },
                                                      },
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                        if: Expr {
                                          raw: {
                                            is_nonempty: Expr {
                                              raw: {
                                                var: "match",
                                              },
                                            },
                                          },
                                        },
                                        then: Expr {
                                          raw: {
                                            get: Expr {
                                              raw: {
                                                var: "match",
                                              },
                                            },
                                          },
                                        },
                                      },
                                    },
                                    let: [
                                      {
                                        match: Expr {
                                          raw: {
                                            match: Expr {
                                              raw: {
                                                index: "users_discord_id",
                                              },
                                            },
                                            terms: Expr {
                                              raw: {
                                                var: "userId",
                                              },
                                            },
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                              {
                                weekend: Expr {
                                  raw: {
                                    var: "weekend",
                                  },
                                },
                              },
                            ],
                          },
                        },
                        lambda: Expr {
                          raw: [
                            "userId",
                            "weekend",
                          ],
                        },
                      },
                    },
                  },
                },
                name: "add_vote_to_user",
              },
            },
          },
        },
      },
      if: Expr {
        raw: {
          exists: Expr {
            raw: {
              function: "add_vote_to_user",
            },
          },
        },
      },
      then: Expr {
        raw: {
          params: Expr {
            raw: {
              object: {
                body: Expr {
                  raw: {
                    query: Expr {
                      raw: {
                        expr: Expr {
                          raw: {
                            in: Expr {
                              raw: {
                                in: Expr {
                                  raw: {
                                    object: {
                                      ok: true,
                                    },
                                  },
                                },
                                let: [
                                  {
                                    user: Expr {
                                      raw: {
                                        params: Expr {
                                          raw: {
                                            object: {
                                              data: Expr {
                                                raw: {
                                                  object: {
                                                    availableVotes: Expr {
                                                      raw: {
                                                        add: Expr {
                                                          raw: [
                                                            Expr {
                                                              raw: {
                                                                default: 0,
                                                                from: Expr {
                                                                  raw: {
                                                                    var: "user",
                                                                  },
                                                                },
                                                                select: Expr {
                                                                  raw: [
                                                                    "data",
                                                                    "availableVotes",
                                                                  ],
                                                                },
                                                              },
                                                            },
                                                            Expr {
                                                              raw: {
                                                                else: 1,
                                                                if: Expr {
                                                                  raw: {
                                                                    equals: Expr {
                                                                      raw: [
                                                                        Expr {
                                                                          raw: {
                                                                            var: "weekend",
                                                                          },
                                                                        },
                                                                        true,
                                                                      ],
                                                                    },
                                                                  },
                                                                },
                                                                then: 2,
                                                              },
                                                            },
                                                          ],
                                                        },
                                                      },
                                                    },
                                                    lastVote: Expr {
                                                      raw: {
                                                        now: null,
                                                      },
                                                    },
                                                    totalVotes: Expr {
                                                      raw: {
                                                        add: Expr {
                                                          raw: [
                                                            Expr {
                                                              raw: {
                                                                default: 0,
                                                                from: Expr {
                                                                  raw: {
                                                                    var: "user",
                                                                  },
                                                                },
                                                                select: Expr {
                                                                  raw: [
                                                                    "data",
                                                                    "totalVotes",
                                                                  ],
                                                                },
                                                              },
                                                            },
                                                            1,
                                                          ],
                                                        },
                                                      },
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                        update: Expr {
                                          raw: {
                                            from: Expr {
                                              raw: {
                                                var: "user",
                                              },
                                            },
                                            select: "ref",
                                          },
                                        },
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                            let: [
                              {
                                user: Expr {
                                  raw: {
                                    in: Expr {
                                      raw: {
                                        else: Expr {
                                          raw: {
                                            create: Expr {
                                              raw: {
                                                collection: "user",
                                              },
                                            },
                                            params: Expr {
                                              raw: {
                                                object: {
                                                  data: Expr {
                                                    raw: {
                                                      object: {
                                                        badges: Expr {
                                                          raw: [
                                                            Expr {
                                                              raw: {
                                                                id: "357600097731608662",
                                                                ref: Expr {
                                                                  raw: {
                                                                    collection: "badge",
                                                                  },
                                                                },
                                                              },
                                                            },
                                                          ],
                                                        },
                                                        id: Expr {
                                                          raw: {
                                                            var: "userId",
                                                          },
                                                        },
                                                        inventories: Expr {
                                                          raw: [
                                                          ],
                                                        },
                                                      },
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                        if: Expr {
                                          raw: {
                                            is_nonempty: Expr {
                                              raw: {
                                                var: "match",
                                              },
                                            },
                                          },
                                        },
                                        then: Expr {
                                          raw: {
                                            get: Expr {
                                              raw: {
                                                var: "match",
                                              },
                                            },
                                          },
                                        },
                                      },
                                    },
                                    let: [
                                      {
                                        match: Expr {
                                          raw: {
                                            match: Expr {
                                              raw: {
                                                index: "users_discord_id",
                                              },
                                            },
                                            terms: Expr {
                                              raw: {
                                                var: "userId",
                                              },
                                            },
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                              {
                                weekend: Expr {
                                  raw: {
                                    var: "weekend",
                                  },
                                },
                              },
                            ],
                          },
                        },
                        lambda: Expr {
                          raw: [
                            "userId",
                            "weekend",
                          ],
                        },
                      },
                    },
                  },
                },
                name: "add_vote_to_user",
              },
            },
          },
          update: Expr {
            raw: {
              function: "add_vote_to_user",
            },
          },
        },
      },
    },
  },
]
`;
